core/src/androidMain/kotlin/Peripheral.kt

```diff
@@ -219,10 +220,8 @@ public class AndroidPeripheral internal constructor(
 
     private fun closeConnection() {
         _connection?.close()
-        if (_state.value !is State.Disconnected) {
-            _state.value = State.Disconnected()
-        }
         _connection = null
+        _state.update { previous -> previous as? State.Disconnected ?: State.Disconnected() }
```

Comment:
This is clever enough that probably warrants a code comment.

```suggestion
        // Set state to `Disconnected` only if not already `Disconnected`.
        _state.update { previous -> previous as? State.Disconnected ?: State.Disconnected() }
```